src/topology/instances/add_circle.lean

```diff
@@ -286,3 +286,115 @@ local attribute [instance] fact_zero_lt_one
+
+variables (T : ℝ) [hT : fact (0 < T)]
+include hT
+local notation `𝕋` := add_circle T
+
+private lemma Icc_to_circle_kernel {x y : ℝ} (hx : x ∈ Icc 0 T) (hy : y ∈ Icc 0 T) :
+  (x : 𝕋) = (y : 𝕋) ↔ (x = y) ∨ (x = 0 ∧ y = T) ∨ (y = 0 ∧ x = T) :=
+begin
+  simp_rw [quotient_add_group.eq, add_subgroup.mem_zmultiples_iff, zsmul_eq_mul],
+  split,
+  { rintros ⟨n, hn⟩,
+    have : | -x + y| ≤ T,
+    { rw abs_le,
+      exact ⟨by linarith [hx.2, hy.1], by linarith [hx.1, hy.2]⟩, },
+    rw [←hn, abs_mul, abs_of_pos hT.out, mul_le_iff_le_one_left hT.out, ←int.cast_abs,
+      ←int.cast_one, int.cast_le] at this,
+    replace this : |n| = 0 ∨ |n| = 1,
+    { rcases eq_or_lt_of_le this with h|h,
+      { exact or.inr h, },
+      { exact or.inl (by linarith [abs_nonneg n]) } },
+    rw [abs_eq (zero_le_one' ℤ), abs_eq_zero] at this,
+    rcases this with (rfl | rfl | rfl),
+    { rw int.cast_zero at hn,
+      exact or.inl (by linarith), },
+    { rw int.cast_one at hn,
+      exact or.inr (or.inl ⟨by linarith [hx.1, hy.2], by linarith [hx.1, hy.2]⟩) },
+    { rw [int.cast_neg, int.cast_one] at hn,
+      exact or.inr (or.inr ⟨by linarith [hy.1, hx.2], by linarith [hy.1, hx.2]⟩), }, },
+  { rintros (rfl | ⟨rfl, h⟩ | ⟨rfl, h⟩),
+    exacts [ ⟨0, by simp⟩, ⟨1, by {rw h, simp}⟩, ⟨-1, by {rw h, simp}⟩] },
+end
+
+variables {T}
+
+/-- The equivalence relation on `Icc 0 T` which identifies `0` and `T`. -/
+private def S : setoid (Icc 0 T) :=
+{ r := λ x y, (x.1 = y.1) ∨ (x.1 = 0 ∧ y.1 = T) ∨ (y.1 = 0 ∧ x.1 = T),
+  iseqv := ⟨(λ x, by tauto), (λ x y hxy, by tauto),
+    (λ x y z hxy hyz, (Icc_to_circle_kernel T x.2 z.2).mp
+    (((Icc_to_circle_kernel T x.2 y.2).mpr hxy).trans
+    ((Icc_to_circle_kernel T y.2 z.2).mpr hyz)))⟩ }
+
+local attribute [instance] S
+
+private lemma Icc_quot_welldef (a b : Icc 0 T) (hab : a ≈ b) : (a : 𝕋) = (b : 𝕋) :=
+(Icc_to_circle_kernel T a.2 b.2).mpr hab
+
+private lemma Icc_quot_bijective : bijective (quotient.lift coe $ @Icc_quot_welldef T hT) :=
+begin
+  refine ⟨λ x y hxy, _, λ x, _⟩,
+  { induction x using quotient.induction_on,
+    induction y using quotient.induction_on,
+    rw [quotient.lift_mk, quotient.lift_mk, coe_coe, coe_coe] at hxy,
+    simpa [quotient.eq] using (Icc_to_circle_kernel T x.2 y.2).mp hxy },
+  { obtain ⟨y, hy⟩ := (by { rw add_circle.coe_image_Icc_eq T, tauto } : x ∈ coe '' Icc 0 T),
+    exact ⟨quotient.mk ⟨y, hy.1⟩, hy.2⟩ },
+end
+
+/-- The natural map from `[0, T]` with endpoints identified to `ℝ / ℤ • T`. This definition
+is private since it is subsumed by `Icc_circle_homeo` later. -/
+private def Icc_circle_equiv : equiv (quotient S) 𝕋 :=
+{ to_fun    := quotient.lift coe $ @Icc_quot_welldef T hT,
+  inv_fun   := classical.some (bijective_iff_has_inverse.mp Icc_quot_bijective),
+  left_inv  := (classical.some_spec $ bijective_iff_has_inverse.mp Icc_quot_bijective).1,
+  right_inv := (classical.some_spec $ bijective_iff_has_inverse.mp Icc_quot_bijective).2 }
```

Comment:
It's longer, but I think it's better to choose an explicit inverse here,
```suggestion
private def Icc_circle_equiv : equiv (quotient S) 𝕋 :=
{ to_fun    := λ x, quotient.lift_on' x coe $ @Icc_quot_welldef T hT,
  inv_fun   := λ x, quotient.mk' $ subtype.map id Ico_subset_Icc_self (add_circle.equiv_Ico _ x),
  left_inv  := quotient.ind' $ subtype.rec $ by exact λ x hx, quotient.sound' $ begin
    dsimp [setoid.r, add_circle.equiv_Ico],
    obtain rfl | hx' := eq_or_ne T x,
    { refine (or.inr $ or.inl $ ⟨_, rfl⟩),
      convert to_Ico_mod_apply_right (0 : ℝ) hT.out,
      rw zero_add },
    { refine or.inl _,
      rw to_Ico_mod_eq_self,
      rw zero_add,
      exact ⟨hx.1, lt_of_le_of_ne' hx.2 hx'⟩ }
  end,
  right_inv := sorry }
```

Reply:
Could you elaborate on why you consider this better? Even with the "sorry" that's four times as long as the existing code to obtain the exact same result.

Reply:
The advantage of this spelling is that you can prove that `Icc_circle_equiv.symm ↑x = quotient.mk (to_Ico_mod _ _)` or similar, and the proof is true by `rfl`.

The other advantage is that on `rat` you can use `#eval` to directly compute the equivalence.

Reply:
> The advantage of this spelling is that you can prove that `Icc_circle_equiv.symm ↑x = quotient.mk (to_Ico_mod _ _)` or similar, and the proof is true by `rfl`.

I don't think that's a very persuasive argument since this statement is only four lines to prove with the current definitions:
```
lemma thing_eric_wants (x : 𝕜) : (add_circle.Icc_circle_equiv p).symm ↑x = 
  quotient.mk (⟨to_Ico_mod 0 hp.out x, mem_of_mem_of_subset (to_Ico_mod_mem_Ico' _ _) Ico_subset_Icc_self⟩) := 
begin
  rw [equiv.apply_eq_iff_eq_symm_apply, equiv.symm_symm],
  dsimp [add_circle.Icc_circle_equiv], 
  rw [←sub_eq_zero, ←quotient_add_group.coe_sub, quotient_add_group.eq_zero_iff,
    to_Ico_mod, ←sub_sub, sub_self, zero_sub, ←neg_zsmul],
  exact ⟨-to_Ico_div 0 hp.out x, by refl⟩,
end
```